language/tools/move-mv-llvm-compiler/src/stackless/module_context.rs

```diff
@@ -622,4 +626,216 @@ impl<'mm, 'up> ModuleContext<'mm, 'up> {
+     * corresponding data structures. The function `deserialize`
+     * returns a triple consiting of
+     * - instruction_data -- a byte array,
+     * - program_id -- SolanaPubkey, and
+     * - accounts -- a vector of SolanaAccountInfo items.
+     *
+     * To select one from possibly several entry functions defined in
+     * the module, the entrypoint function expects the name of the
+     * requested entry function to be passed in instruction_data byte
+     * array. The logic in solana entrypoint iteratively compares the
+     * string slice passed in instruction_data to every entry function
+     * symbol of the module. Once a matching entry function is found,
+     * it is called, and its return value is used as the exit code for
+     * the program.
+     */
+    fn emit_solana_entrypoint(&mut self) {
+        let entry_functions: Vec<_> = self
+            .env
+            .get_functions()
+            .filter(|fn_env| fn_env.is_entry())
+            .collect();
+
+        // Do not generate solana entrypoint if module doesn't contain any entry functions.
+        if entry_functions.is_empty() {
+            return;
+        }
+
+        let deserialize = String::from("deserialize");
+        let ll_sret = self.llvm_cx.get_anonymous_struct_type(&[
+            self.llvm_cx
+                .get_anonymous_struct_type(&[self.llvm_cx.ptr_type(), self.llvm_cx.int_type(64)]),
+            self.llvm_cx.ptr_type(),
+            self.llvm_cx.get_anonymous_struct_type(&[
+                self.llvm_cx.ptr_type(),
+                self.llvm_cx.int_type(64),
+                self.llvm_cx.int_type(64),
+            ]),
+        ]);
+        let ll_fn_deserialize = {
+            let ll_fnty = {
+                let ll_rty = self.llvm_cx.void_type();
+                let sret_param = self.llvm_cx.ptr_type();
+                let ll_param_tys = vec![sret_param, self.llvm_cx.ptr_type()];
+                llvm::FunctionType::new(ll_rty, &ll_param_tys)
+            };
+            let ll_fn = self.llvm_module.add_function(&deserialize, ll_fnty);
+            self.llvm_module
+                .add_type_attribute(ll_fn, 1, "sret", ll_sret);
+            ll_fn
+        };
+        ll_fn_deserialize
+            .as_gv()
+            .set_linkage(llvm::LLVMLinkage::LLVMExternalLinkage);
+        let ret_ty = self.llvm_cx.int_type(1);
+        let len_ty = self.llvm_cx.int_type(64);
+        let ptr_ty = self.llvm_cx.ptr_type();
+        let ll_param_tys = &[ptr_ty, len_ty, ptr_ty, len_ty];
+        let ll_fnty = llvm::FunctionType::new(ret_ty, ll_param_tys);
+        let attrs = vec![
+            (1, "readonly", None),
+            (1, "nonnull", None),
+            (3, "readonly", None),
+            (3, "nonnull", None),
+        ];
+        let ll_fn_str_cmp = self.llvm_module.add_function("move_rt_str_cmp_eq", ll_fnty);
+        self.llvm_module.add_attributes(ll_fn_str_cmp, &attrs);
```

Comment:
Ideally, setting up and calling these new runtime functions should be handled the same way as the other existing ones. See `get_runtime_function`, `get_runtime_function_by_name`, enum `RtCall`, and `emit_rtcall_*`. This will also aid readability and make the code more concise.

Reply:
If you mean the list I mentioned above (`get_runtime_functioon`, etc), then I'd say not now (orthogonal). I have done some recent refactoring of the overall `translate.rs` and moving those to `ModuleContext` is probably the right thing to do (`FunctionContext` hasn't been touched yet and is huge).

Reply:
Then I'm not sure what your suggestion 

> these new runtime functions should be handled the same way as the other existing ones. See get_runtime_function, get_runtime_function_by_name, enum RtCall, and emit_rtcall_*. 

entails. Should I define similar methods `get_runtime_function` et al. in ModuleContext as well?

Reply:
What I mean is to just follow the existing idiom for runtime functions. For example, in `translate.rs`:
```
pub enum RtCall {
    Abort(mast::TempIndex),
    VecDestroy(mast::TempIndex, mty::Type),
    VecCopy(llvm::AnyValue, llvm::AnyValue, mty::Type),
    VecCmpEq(llvm::AnyValue, llvm::AnyValue, mty::Type),
    VecEmpty(mty::Type),
    StructCmpEq(llvm::AnyValue, llvm::AnyValue, mty::Type),
}
```
We'd add the new runtime functions there. And likewise, add cases for the other locations:
```
    fn get_runtime_function(&self, rtcall: &RtCall) -> llvm::Function {
        let name = match rtcall {
            RtCall::Abort(..) => "abort",
            RtCall::VecDestroy(..) => "vec_destroy",
            RtCall::VecCopy(..) => "vec_copy",
            RtCall::VecCmpEq(..) => "vec_cmp_eq",
            RtCall::VecEmpty(..) => "vec_empty",
            RtCall::StructCmpEq(..) => "struct_cmp_eq",
        };
        self.get_runtime_function_by_name(name)
    }
```
And so forth for the others in the list. This factors out the runtime function declaration and calling to the various routines mentioned.
```

Reply:
Right, but this logic exists in FunctionContext currently, and not available to emit_solana_entrypoint, because the latter is a method in ModuleContext. Hence, my question was should the code that you pointed out (from translate.rs) be moved into ModuleContext?

Reply:
If you're already refactoring this and the methods for handling run-time functions will be moved into ModuleContext, I'll wait for your changes, rebase on that, and will use the rt handling functions to call rt_abort etc. I'll also add the new rt methods (cmp_str, deserialize) to the RtCall enum.  In the meantime I can also work on the entry function parameters (the missing part in this PR), so no rush in making those refactorings.

Reply:
SGTM (I'll commit a minimal refactor soon with just the rt-related factored out).